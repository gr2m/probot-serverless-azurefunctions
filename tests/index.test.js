const { serverless } = require('../')
const crypto = require('crypto')

const getEvent = (body = {installation: { id: 1 }}) => {

    const secret = 'iamasecret'

    process.env.WEBHOOK_SECRET = secret
    
    // Create Signature
    let hmac = crypto.createHmac('sha1', secret);
    hmac.update(JSON.stringify(body));
    const hash = hmac.digest('hex');

    return {
      body,
      headers: {
        'X-Github-Event': 'issues',
        'x-Github-Delivery': 123,
        'X-Hub-Signature': `sha1=${hash}`
      }
    }
}

describe('probot-serverless-azurefunctions', () => {
  let spy, handler, context

  beforeEach(() => {
    context = { done: jest.fn(), log: console.log }
    context.log = jest.fn()
    process.env = { }
    spy = jest.fn()
    handler = serverless(async app => {
      app.auth = () => Promise.resolve({})
      app.on('issues', spy)
    })
  })

  it('responds with the homepage', async () => {
    const event = { method: 'GET', path: '/probot' }
    await handler(context, event)
    expect(context.done).toHaveBeenCalled()
    expect(context.res.body).toMatchSnapshot()
    expect(context.log).toMatchSnapshot()
  })

  it('calls the event handler', async () => {
    const event = getEvent()

    await handler(context, event)
    expect(context.done).toHaveBeenCalled()
    expect(spy).toHaveBeenCalled()
  })

  it('responds with a 400 error when body is null', async () => {
    const event = getEvent(null)

    await handler(context, event)
    expect(context.res).toEqual(expect.objectContaining({status: 400}))
    expect(context.done).toHaveBeenCalled()
    expect(spy).not.toHaveBeenCalled()
  })

  it('responds with 400 when payload has been changed', async () => {
    const secret = "iamasecret"

    const body = {
      installation: { id: 1 }
    }

    // Create Signature
    let hmac = crypto.createHmac('sha1', secret);
    hmac.update(JSON.stringify(body));
    const hash = hmac.digest('hex');

    const event = {
      body: {
        installation: { id: 123123 } // different to what 'github' would have sent
      },
      headers: {
        'X-Github-Event': 'issues',
        'x-Github-Delivery': 123,
        'X-Hub-Signature': `sha1=${hash}`
      }
    }

    process.env.WEBHOOK_SECRET = secret

    await handler(context, event)
    expect(context.res).toEqual(expect.objectContaining({status: 400}))
    expect(context.done).toHaveBeenCalled()
    expect(spy).not.toHaveBeenCalled()
    expect(context.log).toMatchSnapshot()
  })
})