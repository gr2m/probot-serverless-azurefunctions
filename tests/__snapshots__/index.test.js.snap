// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`probot-serverless-azurefunctions responds with 400 when payload has been changed 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Received event: issues",
    ],
    Array [
      "Event executed with status 400 and output of: ",
      Array [
        Object {
          "event": Object {
            "id": 123,
            "name": "issues",
            "payload": Object {
              "installation": Object {
                "id": 123123,
              },
            },
            "signature": "sha1=11daab007b2f76317fa026fdb6f3d66337a9d701",
          },
          "status": 400,
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`probot-serverless-azurefunctions responds with 400 when payload has been changed 2`] = `[MockFunction]`;

exports[`probot-serverless-azurefunctions responds with the homepage 1`] = `
"
<!DOCTYPE html>
<html lang=\\"en\\" class=\\"height-full\\">
  <head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
    <title>Your App | built with Probot</title>
    <link rel=\\"stylesheet\\" href=\\"https://probot.github.io/assets/css/index.css?d=1521919566\\">
  </head>
  <body class=\\"height-full bg-gray-light\\">
    <div class=\\"d-flex flex-column flex-justify-center flex-items-center text-center height-full\\">
      <img src=\\"https://probot.github.io/assets/logo.png\\" alt=\\"Probot Logo\\" width=\\"100\\" class=\\"mb-6\\">
      <div class=\\"box-shadow rounded-2 border p-6 bg-white\\">
        <h1>
          Welcome to your Probot App!
        </h1>
        <p>This bot was built using <a href=\\"https://github.com/probot/probot\\">Probot</a>, a framework for building GitHub Apps.</p>
      </div>
      <div class=\\"mt-4\\">
        <h4 class=\\"alt-h4 text-gray-light\\">Need help?</h4>
        <div class=\\"d-flex flex-justify-center mt-2\\">
          <a href=\\"https://probot.github.io/docs/\\" class=\\"btn btn-outline mr-2\\">Documentation</a>
          <a href=\\"https://probot-slackin.herokuapp.com/\\" class=\\"btn btn-outline\\">Chat on Slack</a>
        </div>
      </div>
    </div>
  </body>
</html>
"
`;

exports[`probot-serverless-azurefunctions responds with the homepage 2`] = `[MockFunction]`;

exports[`probot-serverless-azurefunctions responds with the homepage 3`] = `[MockFunction]`;
